documentation/testing_plans.md

```diff
@@ -0,0 +1,783 @@
+# Testing plans
+
+Bolt ships with a library of helpers, named `BoltSpec`, intended to be used for
+writing unit tests for plans with the `RSpec` testing tool. Before writing unit
+tests for plans, get familiar with [RSpec](https://rspec.info/).
+
+## Set up the test environment
+
+- Install the [Puppet Development Kit
+  (PDK)](https://puppet.com/docs/pdk/latest/pdk_install.html), which is used to
+  develop Puppet modules and provides integrated testing tools.
+
+- Create a [new
+  module](https://puppet.com/docs/pdk/latest/pdk_creating_modules.html) using
+  `pdk new module` or [convert an existing
+  module](https://puppet.com/docs/pdk/latest/pdk_converting_modules.html) with
+  `pdk convert`.
+
+- Ensure your module has files at `Gemfile`, `Rakefile`, and
+  `spec/spec_helper.rb`.
+
+- Install the `bundler` gem with Ruby:
+
+  ```shell
+  gem install bundler
+  ```
+
+- Install the module's gem dependencies, which include Bolt and other testing
+  tools:
+
+  ```shell
+  bundle install
+  ```
+
+## Directory and file structure
+
+Tests should be saved to the same module that the plans they are testing are
+saved to.
+
+By convention, tests for a plan are saved to a file with the name `<PLAN
+NAME>_spec.rb` in the module's `spec/plans` directory.
+
+```shell
+my_module/
+├── Gemfile
+├── metadata.json
+├── plans/
+│   └── my_plan.pp
+├── Rakefile
+└── spec/
+    ├── plans/
+    │   └── my_plan_spec.rb
+    └── spec_helper.rb
+```
+
+
+## Include BoltSpec functions
+
+To use the `BoltSpec` library functions, you need to include them in either the
+`spec_helper.rb` file or in the individual test files. You also need to
+configure Puppet and Bolt for the testing environment.
+
+### Include BoltSpec in spec_helper.rb
+
+To expose the `BoltSpec` library functions and configure Puppet and Bolt for
+all tests in the `spec/` directory, add the following lines to the `spec_helper.rb`
+file:
+
+```ruby
+# spec/spec_helper.rb
+
+# Load the BoltSpec library
+require 'bolt_spec/plans'
+
+# Include the BoltSpec library functions
+include BoltSpec::Plans
+
+# Configure Puppet and Bolt for testing
+BoltSpec::Plans.init
+
+# Additional spec_helper configuration . . .
+```
+
+### Include BoltSpec in test files
+
+To expose the `BoltSpec` helper functions and configure Puppet and Bolt for
+a single test file, add the following lines to the test file:
```

Comment:
There's a comment in BoltSpec that says running `BoltSpec::Plans.init` probably isn't safe to do in modules that also test manifest code. I _think_ that only applies if you're doing that in `spec_helper.rb`, since it would apply it to all tests.

We could add a line that suggests doing it from `spec_helper.rb` unless you're also testing manifest code?

Reply:
Cool - yeah I language around preferring spec_helper unless you're testing manifests makes sense